### PR TITLE
Issue/jces 2122 fix throttled cache

### PR DESCRIPTION
**Pull request by wyrzyk**

### CODE REVIEW

File: src/test/java/com/atlassian/db/replica/api/ThrottledCacheTest.java [hunk 1]
```diff
@@ -144,4 +157,16 @@ public Instant instant() {
             return instant;
         }
     }
+
+    private final class Work {
```
Comment (afaruga-atlassian) [inline src/test/java/com/atlassian/db/replica/api/ThrottledCacheTest.java:161]: `WaitingWork`

File: src/test/java/com/atlassian/db/replica/api/ThrottledCacheTest.java [hunk 2]
```diff
@@ -94,28 +95,40 @@ public void shouldntSupplierBlockTheCacheForever() throws InterruptedException {
...
+        ThrottledCache<Long> cache = new ThrottledCache<>(clock, ofMillis(1500));
+
+        final Work firstThreadWork = asyncPutWithSlowSupplier(cache, 1);
+        clock.tick();
+        clock.tick();
+        asyncPutWithSlowSupplier(cache, 2);
+        firstThreadWork.finish();
+
+        cache.get(() -> {
+            throw new ConcurrentModificationException("Concurrent invalidation is not allowed");
```
Comment (afaruga-atlassian) [inline src/test/java/com/atlassian/db/replica/api/ThrottledCacheTest.java:109]: I'm missing assertion in this test.

File: CHANGELOG.md [hunk 3]
```diff
@@ -22,7 +22,13 @@ Dropping a requirement of a major version of a dependency is a new contract.
 
 
 ## [Unreleased]
-[Unreleased]: https://github.com/atlassian-labs/db-replica/compare/release-2.0.0...master
+[Unreleased]: https://github.com/atlassian-labs/db-replica/compare/release-2.1.0...master
+
+### Fixed
+- Fix `Throttled Cache` lock release mechanism
```
Comment (afaruga-atlassian) [inline CHANGELOG.md:28]: No mention about `LockBasedThrottledCache`

File: src/main/java/com/atlassian/db/replica/internal/LockBasedThrottledCache.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.atlassian.db.replica.internal.util.ThreadSafe;
+import com.atlassian.db.replica.spi.SuppliedCache;
+
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+@ThreadSafe
+public final class LockBasedThrottledCache<T> implements SuppliedCache<T> {
```
Comment (afaruga-atlassian) [inline src/main/java/com/atlassian/db/replica/internal/LockBasedThrottledCache.java:11]: Should it be hidden by `compatibleWithPreviousVerions` 🤔

File: CHANGELOG.md (hunk 3 above)
Comment (wyrzyk) [reply CHANGELOG.md:28]: It's not part of the API

File: src/main/java/com/atlassian/db/replica/internal/LockBasedThrottledCache.java (hunk 4 above)
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/LockBasedThrottledCache.java:11]: We use `compatibleWithPreviousVerions` for the cases when you can't opt-out from the change via API.

Comment (afaruga-atlassian) [reply src/main/java/com/atlassian/db/replica/internal/LockBasedThrottledCache.java:11]: Ach, yes, this cache is a part of API. OK.

File: src/main/java/com/atlassian/db/replica/api/ThrottledCache.java [hunk 5]
```diff
@@ -17,58 +18,76 @@
...
-    private final AtomicReference<Instant> lock = new AtomicReference<>();
-    private final Clock clock;
-    private final Duration timeout;
-    private T value = null;
+    private final SuppliedCache<T> delegate;
 
     public ThrottledCache(Clock clock, Duration timeout) {
-        this.clock = clock;
-        this.timeout = timeout;
+        this.delegate = new ThrottledCacheWithTimeout<>(clock, timeout);
```
Comment (afaruga-atlassian) [inline src/main/java/com/atlassian/db/replica/api/ThrottledCache.java:24]: ahh, this is how you hide it, and make it backward compatible. 
IMHO visibility of this change from a customer perspective is low.
Maybe java doc? `public ThrottledCache(Clock clock, Duration timeout) {`

**Review by afaruga-atlassian — APPROVED** [review]

**Pull request merged**